.codecov.yml

```diff
@@ -41,10 +41,11 @@ parsers:
       macro: no
 
 ignore:
-  - "bfintrinsics/**"
-  - "bfm/ioctl/arch/**"
-  - "bfvmm/include/test/**"
-  - "bfdriver/**"
+  - "bfintrinsics/**"           # Not useful to unit test
+  - "bfvmm/include/test/**"     # Don't calculate coverage on unit test code
+  - "bfdriver/**"               # Not possible to unit test (OS drivers)
+  - "bfvmm/include/sdk/**"      # Covered with integration tests
```

Comment:
I am fine with this so long as this is header only and only wraps around the existing APIs.

Reply:
I don't want to make these header-only for a couple reasons:

1) The point of the SDK functions is to define an interface that hides implementation details. Seperating into a header and source is a logical way to accomplish that separation.

2) When we put implementations in a header file, the files end up looking like [this](https://github.com/Bareflank/hypervisor/blob/e3b641225ead1fee2b2c3bdeb2c32d84e4265f44/bfvmm/include/hve/arch/intel_x64/ept/helpers.h). In my opinion, there's a lot of unnecessary "C++ clutter" in there that distracts from learning how to use the interface.

Reply:
The problem is, the source approach adds symbols to the runtime linking phase and it prevents inlining, so the compiler cannot optimize out your code. Basically, you are adding additional function calls to every API for the sole purpose of reducing clutter. Also, its not a C++ thing, inlining is done in C as well. 
https://www.kernel.org/doc/local/inline.html

Reply:
Im somewhat torn here, so I'm going to ask some questions first:
> The point of the SDK functions is to define an interface that hides implementation details

@JaredWright  what is the problem(s) that hiding the implementation solves?

> so the compiler cannot optimize out your code

@rianquinn do you have any numbers for this? Are we sure that the lack of optimization has a negative impact for Bareflank specifically?

Reply:
I don't have any numbers for this specific use-case, but wikipedia claims 10-20%: https://en.wikipedia.org/wiki/Inline_expansion#Effect_on_performance

More importantly, the impact would be a lot higher for this specific case because it's not just about expansion. In this case, the compiler will literally remove the function call wrapper as it is not needed. If you place this code in source, the compiler is forced to mark the code as global (even in the static compilation case), and will be present, meaning these APIs will be adding an additional function call for every API. 

Like I said before, this is not a C++ thing. Here is one of a million examples in the Linux kernel:
https://github.com/torvalds/linux/blob/master/arch/x86/include/asm/apic.h#L102

We should be moving towards inlining, not away from it. Especially if the function is designed to wrap up code that we would normally write ourselves. C++ in a lot of ways is a header-only language. Meta-programming requires it. Just look at libc++ and how it is implemented. The only source files are for things that absolutely need it. I am all for making our APIs more approachable, but not at the expense of turning C++ into C, or hurting performance. 

Clang 5 BTW added modules support. For the VMM at least, we could always add modules support to improve compilation performance. But even with that specific issue, the ship has sailed as the intrinsics are the thing that needs that. These wrappers are just noise.

Reply:
One way to solve this BTW, is to do what libc++ does:
https://github.com/Bareflank/libcxx/blob/master/include/memory#L26

I might be misunderstanding the objection, but it seems to me that the biggest issue @JaredWright has is the way the code looks. To solve this, we could always do this:
https://github.com/Bareflank/libcxx/blob/master/include/memory#L26

Basicaly, at the start of the file in a comment block, you document all of the functions for Doygen without the implementation, and then the implementation follows without comments. This way, if you are a user, you simply read the comment section like you would an unimplemented header file, but you get the performance benefits of inlining.

Reply:
@connojd To clarify your question:

> what is the problem(s) that hiding the implementation solves?

The problem that I foresee hiding implementation details (i.e. declaring abstract interfaces) solving is instability between extensions and the base hypervisor. I'm hypothesizing that defining purely abstract interfaces will encourage both us (the library developers) and users (extension writers) to think more in terms of abstract interactions and end user needs, and less in terms of how that need is accomplished at a given moment in time. As @rianquinn points out, lots of things in C++ (including the standard library) are *implemented* as header only. The keyword there is implemented, because even those header files are conforming to a separate abstract specification (the ISO C++ standard). Right now, Bareflank doesn't have this concept of "abstract specification", so I'm asserting that separating headers from implementations will help start us down that path. I have no way to prove if this will be successful, but I'd like to give it a try.

> Are we sure that the lack of optimization has a negative impact for Bareflank specifically?

Agreed on this point. I'm not willing to talk about performance and optimizations at this level (removing one function call) until we have a way to measure those impacts.